rero_ils/modules/documents/commons/identifiers.py

```diff
@@ -0,0 +1,281 @@
+    LOCAL = 'bf:Local'
+    MATRIX_NUMBER = 'bf:MatrixNumber'
+    MUSIC_DISTRIBUTOR_NUMBER = 'bf:MusicDistributorNumber'
+    MUSIC_PLATE = 'bf:MusicPlate'
+    MUSIC_PUBLISHER_NUMBER = 'bf:MusicPublisherNumber'
+    PUBLISHER_NUMBER = 'bf:PublisherNumber'
+    UPC = 'bf:Upc'
+    URN = 'bf:Urn'
+    VIDEO_RECORDING_NUMBER = 'bf:VideoRecordingNumber'
+    URI = 'uri'
+
+
+class IdentifierStatus:
+    """Status of identifier."""
+
+    UNDEFINED = None
+    INVALID = 'invalid'
+    CANCELLED = 'cancelled'
+    INVALID_OR_CANCELLED = 'invalid or cancelled'
+
+
+# =============================================================================
+#     IDENTIFIER
+# =============================================================================
+#    * `Identifier` class is the most generic class for a document identifier.
+#    * `ISBNIdentifier` class represent any ISBN identifier (isbn-10, isbn-13)
+#    * `EANIdentifier` class to represent an isbn without any hyphens
+# =============================================================================
+DocIdentifier = TypeVar('DocIdentifier')
+
+
+@dataclass(repr=False)
+class Identifier:
+    """Generic document identifier representation."""
+
+    value: str
+    type: Optional[str] = None
+    note: Optional[str] = None
+    qualifier: Optional[str] = None
+    source: Optional[str] = None
+    status: str = field(default=IdentifierStatus.UNDEFINED)
+
+    def __post_init__(self):
+        """Post initialization dataclass magic function."""
+        if self.status == IdentifierStatus.UNDEFINED and not self.is_valid():
+            self.status = IdentifierStatus.INVALID
+        if hasattr(self, '__type__'):
+            self.type = self.__type__
+        if not self.type:
+            raise InvalidIdentifierException("'type' is a required property.")
+
+    def __hash__(self):
+        """Get an hash value for this identifier."""
+        return hash(self.type + self.value)
+
+    def __str__(self) -> str:
+        """Get simple string representation of the identifier."""
+        return self.normalize()
+
+    def normalize(self) -> str:
+        """Get the normalized value of the identifier."""
+        return self.value
+
+    def is_valid(self) -> bool:
+        """Check if the identifier is valid."""
+        # As we are in a generic class, all identifier are valid.
+        return True
+
+    def validate(self) -> None:
+        """Validate the identifier."""
+        # same as `is_valid` but raise an exception if the identifier isn't
+        # valid
+        if not self.is_valid():
+            raise InvalidIdentifierException()
+
+    def render(self, render_class=None) -> str:
+        """Render the identifier.
+
+        :param render_class: a rendering class to use to get the identifier
+                             representation.
+        :returns the string representation of the identifier.
+        """
+        render_class = render_class or DefaultIdentifierRenderer()
```

Comment:
I would use a class variable for this as done for the model_cls in the record class.